### PR TITLE
MSP432 SPI driver

### PR DESCRIPTION
**Pull request by LosDeiblos**
### Pull Request Overview

This PR adds support for SPI on the MSP432. The SPI related registers in the USCI modules are identical, but the usage of certain bits is different to UART and I2C mode. As a solution I simply cast the concerned registers to UCSPIx-registers instead of using the UCAx/UCBx-registers. With this trick the references of UCAx/UCBx-registers can be passed to the SPI-constructor.

### Testing Strategy

Tested on the evaluation board using a the spi_controller_transfer example as application.

### TODO or Help Wanted

* Is the approach of casting the UCAx/UCBx-registers to a common type ok/fine? Or should I rewrite the driver to use a different approach?
* I tested the SPI driver also using a [Waveshare 2.7inch E-Ink display](https://www.waveshare.com/w/upload/2/2d/2.7inch-e-paper-Specification.pdf) as a slave. Is there an interest for a driver-support for this E-Ink display? If yes, I could create a new PR for it.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: boards/msp_exp432p401r/Makefile [hunk 1]
```diff
@@ -19,7 +19,7 @@ flash-debug: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/debug/$(PLATFORM).elf
 	$(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $<; verify_image $<; reset; shutdown"
 
 .PHONY: flash
-flash: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/release/$(PLATFORM).elf
+flash: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/release/$(PLATFORM).bin
```
Comment (bradjc) [inline boards/msp_exp432p401r/Makefile:22]: Is this change needed?

File: chips/msp432/src/spi.rs [hunk 2]
```diff
@@ -0,0 +1,469 @@
+use crate::dma;
```
Comment (bradjc) [inline chips/msp432/src/spi.rs:1]: Please add `//!` comment.

**Review by bradjc — APPROVED** [review]
This looks pretty straightforward to me. I do have a question about the Makefile change.

**Comment by bradjc** [comment]
>     * I tested the SPI driver also using a [Waveshare 2.7inch E-Ink display](https://www.waveshare.com/w/upload/2/2d/2.7inch-e-paper-Specification.pdf) as a slave. Is there an interest for a driver-support for this E-Ink display? If yes, I could create a new PR for it.

Yes! I was thinking about wanting an e-ink display last week, in fact.

File: chips/msp432/src/usci.rs [hunk 3]
```diff
@@ -102,6 +102,147 @@ register_structs! {
...
+            /// Enabled. eUSCI logic held in reset state
+            Enabled = 1
+        ],
+        /// STE mode select in master mode
+        UCSTEM OFFSET(1) NUMBITS(1) [
+            /// STE pin is used to prevent conflicts with other masters
+            PreventConflicts = 0,
+            /// STE pin is used to generate the enable signal for a 4-wire slave
+            ChipSelect = 1
+        ],
```
Comment (hudson-ayers) [inline chips/msp432/src/usci.rs:119]: I am trying to verify the safety of the transmute used in `spi.rs`. I am confused, however, because this bitfield is clearly different than the bit described at offset 1 in `UCBxCTLW0`. I suspect it is because I do not understand how these registers work on the MSP432. Is the idea that these registers have a totally different meaning depending on if the module is configured for SPI vs. if it is configured for I2C? And thus the bitfield descriptions are different in that case?

File: boards/msp_exp432p401r/Makefile (hunk 1 above)
Comment (LosDeiblos) [reply boards/msp_exp432p401r/Makefile:22]: During testing I noticed that flashing the kernel using `make flash` overwrites the .apps section, which requires to also flash the application again. This is probably the case, because the elf file contains a 0-byte as a placeholder in section .apps. As far as I understand the .bin file doesn't contain this placeholder. Therefore I changed the Makefile to flash the binary instead of the elf file, which allows to flash the kernel independently from the application.

File: chips/msp432/src/spi.rs (hunk 2 above)
Comment (LosDeiblos) [reply chips/msp432/src/spi.rs:8]: Added the comments now :)

**Comment by LosDeiblos** [comment]
> > ```
> > * I tested the SPI driver also using a [Waveshare 2.7inch E-Ink display](https://www.waveshare.com/w/upload/2/2d/2.7inch-e-paper-Specification.pdf) as a slave. Is there an interest for a driver-support for this E-Ink display? If yes, I could create a new PR for it.
> > ```
> 
> Yes! I was thinking about wanting an e-ink display last week, in fact.

Great! Then I will add it hopefully very soon. I assume that I should use the screen trait for it?

File: chips/msp432/src/spi.rs [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+/// references of USCI_A and USCI_B modules to the Spi-constructor.
+pub trait UsciSpiRef {
+    fn ctlw0(&self) -> &ReadWrite<u16, UCSPIxCTLW0::Register>;
+    fn brw(&self) -> &ReadWrite<u16>;
+    fn statw(&self) -> &ReadWrite<u16, UCSPIxSTATW::Register>;
+    fn rxbuf(&self) -> &ReadOnly<u16>;
+    fn txbuf(&self) -> &ReadWrite<u16>;
+    fn ie(&self) -> &ReadWrite<u16, UCSPIxIE::Register>;
+    fn ifg(&self) -> &ReadWrite<u16, UCSPIxIFG::Register>;
+}
```
Comment (phil-levis) [inline chips/msp432/src/spi.rs:32]: This seems like a reasonable approach, but could perhaps be explained better. You aren't casting, you are encapsulating. I.e., the two actual bit mappings are different, so you define a common abstraction (`UsciSpiRef`, which should perhaps be `UsciSpiRegisters`), such that the SPI implementation can operate on top of either one.

I would suggest having the Spi type be a type parameter, rather than a `dyn`, though. We've found that this reduces code size and reduces overheads.

File: chips/msp432/src/spi.rs [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    tx_dma_src: u8,
+
+    rx_dma: OptionalCell<&'a dma::DmaChannel<'a>>,
+    pub(crate) rx_dma_chan: usize,
+    rx_dma_src: u8,
+}
+
+impl<'a> Spi<'a> {
+    pub fn new(
+        registers: &'static dyn UsciSpiRef,
```
Comment (phil-levis) [inline chips/msp432/src/spi.rs:189]: I'd suggest parameterizing Spi by the `UsciSpiRef` implementation rather than having a `dyn`.

File: chips/msp432/src/spi.rs (hunk 4 above)
Comment (hudson-ayers) [reply chips/msp432/src/spi.rs:32]: The underlying type is in fact transmuted in the implementation of this trait.

File: chips/msp432/src/spi.rs [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    fn ifg(&self) -> &ReadWrite<u16, UCSPIxIFG::Register>;
+}
+
+impl UsciSpiRef for StaticRef<usci::UsciBRegisters> {
+    fn ctlw0(&self) -> &ReadWrite<u16, usci::UCSPIxCTLW0::Register> {
+        unsafe {
+            &*(&self.ctlw0 as *const ReadWrite<u16, usci::UCBxCTLW0::Register>
+                as *const ReadWrite<u16, usci::UCSPIxCTLW0::Register>)
+        }
+    }
```
Comment (hudson-ayers) [inline chips/msp432/src/spi.rs:40]: This makes me somewhat uncomfortable. I understand the approach and why it works, but this is basically a `transmute` and my experience is that proving a transmute like this will not cause UB is not always as straightforward as it seems.

At the very least I think there should be a `// SAFETY: ` comment above this describing why it is a safe operation. If @jrvanwhy has a chance to weigh in that might also be nice.

**Review by bradjc — CHANGES_REQUESTED** [review]
I had approved, but I didn't look closely enough for the unsafe use. That warrants a closer look because 1) problems from unsafe are often very subtle, 2) unsafe can be difficult to remove later and we want to avoid setting examples that others then copy.

**Comment by lschuermann** [comment]
I'm wondering whether it might be a good idea to have a wrapper around a `tock-registers` register, which can be used to treat a register as having a different `RegisterLongName`? From a safety perspective that should be totally fine, given that on registers we can simply set a raw value directly. Especially with #2517, we should be able to rather seamlessly integrate something like this and then derive the accessor traits accordingly. I'll create a PR once #2517 makes it in.

Depending on how long #2517 will take and whether such an "RegisterAlias" wrapper-type is actually desired, I'd vouch for not using `transmute`, but instead work around this by constructing a `LocalRegisterCopy` to use the `tock-registers` interface and read/write raw values using `get()` / `set()`.

**Comment by phil-levis** [comment]
This has been waiting for 8 months -- closing it.
